mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp

```diff
@@ -1332,6 +1332,120 @@ struct TosaFoldConstantMaximum
   }
 };
 
+template <typename AccumulatorType, typename InputType>
+SmallVector<AccumulatorType> matmul(
+    ShapedType outputType, ElementsAttr matrixA, ElementsAttr matrixB,
+    const std::function<AccumulatorType(const InputType &)> &convertToAccType) {
+
+  auto inputAShape = cast<ShapedType>(matrixA.getType()).getShape();
+  auto inputBShape = cast<ShapedType>(matrixB.getType()).getShape();
+
+  // InputA -> (NHC), InputB -> (NCW)
+  constexpr int64_t batchDim = 0;
+  constexpr int64_t heightDim = 1;
+  constexpr int64_t channelDim = 2;
+  constexpr int64_t widthDim = 2;
+  const auto batchSize = inputAShape[batchDim];
+  const auto channelSize = inputAShape[channelDim];
+  const auto heightSize = inputAShape[heightDim];
+  const auto widthSize = inputBShape[widthDim];
+
+  SmallVector<AccumulatorType> outputValues(outputType.getNumElements());
+  auto matrixAVals = matrixA.getValues<InputType>();
+  auto matrixBVals = matrixB.getValues<InputType>();
+
+  // Output index is always incremented by one, so avoid computing its index for
+  // each iteration.
+  auto indexOut = 0;
+  for (int64_t batch = 0; batch < batchSize; ++batch) {
+    for (int64_t height = 0; height < heightSize; ++height) {
+      for (int64_t width = 0; width < widthSize; ++width, ++indexOut) {
+        AccumulatorType acc = static_cast<AccumulatorType>(0);
+        auto indexA =
+            indexToOffset(inputAShape, {batch, height, /*channel=*/0});
+        auto indexB = indexToOffset(inputBShape, {batch, /*channel=*/0, width});
+        for (int64_t channel = 0; channel < channelSize;
+             ++channel, ++indexA, indexB += widthSize) {
+          acc += (convertToAccType(matrixAVals[indexA]) *
+                  convertToAccType(matrixBVals[indexB]));
+        }
+        outputValues[indexOut] = acc;
+      }
+    }
+  }
+
+  return outputValues;
+}
+
+DenseElementsAttr matmul(ShapedType outputType, ElementsAttr matrixA,
+                         ElementsAttr matrixB) {
+
+  auto baseType = outputType.getElementType();
+  if (auto intType = dyn_cast<IntegerType>(outputType.getElementType())) {
+    // For integer types, accumulate values in int64_t to allow support for i8
+    // and i16 that accumulates with i32 and i48, respectively.
+    auto values = matmul<int64_t, APInt>(
+        outputType, matrixA, matrixB,
+        [&](const APInt &val) { return val.getSExtValue(); });
+
+    // Correct int32_t to the correct output type.
```

Comment:
Does this conversion use the correct saturation technique? Do you have a test case that shows what happens when the computed values are outside of the accumulator type? 
```suggestion
    // Correct int64_t to the correct output type.
```